---
title: "E07 - RNA-seq / 01 - quality control (Analysis of Gene Expression @ UCT Prague)"
author:
  - Jiri Novotny jiri.novotny@img.cas.cz
  - Studuj bioinformatiku! http://studuj.bioinformatiku.cz
institute: "Laboratory of Genomics and Bioinformatics @ Institute of Molecular Genetics of the ASCR"
output:
  rmdformats::readthedown:
    highlight: "kate"
    lightbox: true
    thumbnails: true
    gallery: true
    toc_depth: 4
    self_contained: true
    number_sections: false
    toc_collapsed: false
    df_print: "paged"
date: "`r Sys.Date()`"
---

```{r, child = here("_assets/custom.Rmd"), eval = TRUE}
```

***

Copy, please, these files and directories to your personal directory:

```{bash}
mkdir -p ~/AGE2021/Exercises/E07-RNA_seq/data
cp -r ~/shared/AGE2021/Exercises/E07-RNA_seq/data/experiment ~/AGE2021/Exercises/E07-RNA_seq/data
cp -r ~/shared/AGE2021/Exercises/E07-RNA_seq/01_quality_control ~/AGE2021/Exercises/E07-RNA_seq
```

In this and following exercises, we will be mostly working in shell.
You can send current line or selected code to the terminal in RStudio with `Ctrl + Alt + Enter`.
Now go to the terminal tab at the bottom and change the working directory to the current exercise's directory.
If there isn't the terminal tab, you have to open a new one in `Tools -> Terminal -> New Terminal`.

```{bash}
cd ~/AGE2021/Exercises/E07-RNA_seq/01_quality_control
```

> Unfortunately, in `bash`, there isn't something similar to the R's project-based/`here()` relative navigation,
  so better is to switch working directory and navigate using paths relative to that.

Some commands will take time to finish, so I recommend you to work in a `tmux` session,
as you will be anytime able to exit your SSH connection and later come back to your work.

```{bash}
tmux new -s rnaseq
```

You can detach the `tmux` session with

```{bash}
tmux detach
```

and attach again with

```{bash}
tmux attach -t rnaseq
```

Because Python 2 is the default one on our VMs, we need to install `MultiQC` tool (will be introduced later) for Python 3.
For that purpose we create a new `conda` environment to which we also install additional tools in the following exercise:

```{bash}
conda init bash
conda create -n tools
conda deactivate
conda activate tools
conda install -c bioconda multiqc
```

> To easily distinguish between R and bash code chunks, you can hover over a chunk and the code type will appear on the left side.

> Because `bash` chunks don't share variables between them, you have to run code in the terminal (copy-paste or `Ctrl + Alt + Enter`).
  If you just run a chunk with `bash` code, it will be executed in a separate shell session and, thus, all variables will be forgetted.

***

# Introduction

From [Wikipedia](https://en.wikipedia.org/wiki/RNA-Seq):

> RNA-seq (named as an abbreviation of "RNA sequencing") is a particular technology-based sequencing technique which
uses next-generation sequencing (NGS) to reveal the presence and quantity of RNA in a biological sample at a given moment,
analyzing the continuously changing cellular transcriptome.

![A typical RNA-seq pipeline in which we want to compare two or several experimental conditions.
This pipeline is in principle the same for different research questions, such as differential expression, alternative splicing, mutations (SNPs),
gene fusions, de novo transcriptome assembly etc.
In the first two steps, RNA is isolated and prepared for sequencing by reverse transcription (cDNA) and linker ligation.
After sequencing you obtain short reads of original RNA, and this is where data analysis begins (see next picture).
You usually obtain demultiplexed reads in FASTQ format, i.e. separate file for each sample.
If you are starting directly from data produced by sequencer, you have to demultiplex the reads with Illumina`s tool
[bcl2fastq](https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html).
[Source](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004393)](`r here("E07-RNA_seq/01_quality_control/_rmd_images/rnaseq_pipeline.png")`)

![RNA-seq data analysis workflow. It starts by a technical quality control of reads which tells you if sequencing went technically good.
This is partially also possible to do in BaseSpace, a cloud for Illumina sequencing platform.
The second step is an alignment of reads to reference genome or transcriptome.
This tells us from which genes (or more precisely, from which part of genome) our reads are coming from.
Then we use the information about the alignment to do a quantification, that is, we calculate how many reads are aligned to genes and use
this information to obtain a transcript abundance.
Finally, we use the transcript abundance for differential expression, that is, a relative change in the transcript abundance
(i.e. gene expression) between experimental conditions.
Results from the differential expression can be used for higher-level analysis - for example a functional profiling, in which
we are looking at changes at the level of whole gene sets (e.g. pathways, cellular components, diseases, etc.).
[Source](https://www.1010genome.com/rna-seq/)](`r here("E07-RNA_seq/01_quality_control/_rmd_images/rnaseq_data_analysis_pipeline.png")`)

You may find useful [this](https://github.com/crazyhottommy/RNA-seq-analysis) awesome list of RNA-seq related tutorials and readings.

***

In this part of RNA-seq exercises we will do the technical quality control of an RNA-seq experiment.

Although there are complete pipelines for RNA-seq, such as [nf-core/rnaseq](https://nf-co.re/rnaseq) used by our core facility,
we will go through all the steps manually so that you know the principles.

***

# Our experiment

In this [experiment](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778),
human airway smooth muscle cells were treated with three anti-asthma drugs.
Samples were sequenced using Illumina HiSeq 2000 in paired-end mode, that is,
for each sample there are forward and reverse reads
(they are usually denoted as `1` and `2`, or `left` and `right`, respectively).

This experiment is also used in a great tutorial
[RNA-Seq workflow: gene-level exploratory analysis and differential expression](https://f1000research.com/articles/4-1070/v2)
of which this exercise is largely based on.

Usually, databases such as [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/),
[ArrayExpress](https://www.ebi.ac.uk/arrayexpress/) or [BioStudies](https://www.ebi.ac.uk/biostudies/) will provide you count matrices,
which can be directly imported to R.
You can see [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778) that for this experiment
FPKM matrix is available.

The situation is different for raw reads:

- In [GEO](https://www.ncbi.nlm.nih.gov/geo/), raw reads are stored in [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra)
  and must be retrieved (as FASTQ files) through `fasterq-dump` tool, which will be introduced later.
  - [Here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE174611) is an example study from GEO.
    There is a [link](https://www.ncbi.nlm.nih.gov/sra?term=SRP320304) to its associated samples in SRA.
    On that page you can click on "Send results to Run selector", which will bring you to a
    [SRA page](https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_60a61baed2c2ae3018511a2e&o=acc_s%3Aa&s=SRR14575515,SRR14575528,SRR14575532,SRR14575533,SRR14575535,SRR14575537,SRR14575541,SRR14575542,SRR14575519,SRR14575526,SRR14575539,SRR14575512,SRR14575513,SRR14575516,SRR14575517,SRR14575520,SRR14575521,SRR14575522,SRR14575523,SRR14575524,SRR14575525,SRR14575529,SRR14575531,SRR14575538,SRR14575543,SRR14575514,SRR14575518,SRR14575527,SRR14575530,SRR14575534,SRR14575536,SRR14575540)
    with the samples' raw data.
    Then you have to download the metadata with SRA IDs and use them to retrieve the FASTQ files through `fasterq-dump`.
- In [ArrayExpress](https://www.ebi.ac.uk/arrayexpress/), raw data (FASTQ files) are deposited in
  [European Nucleotide Archive (ENA)](https://www.ebi.ac.uk/ena/browser/home) and could be directly downloaded.
  - [Here](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-9624/) is an example study from ArrayExpress.
    There is a [link](https://www.ebi.ac.uk/ena/browser/view/PRJEB40610) to ENA page with all the samples' raw data in FASTQ format.

Because RNA-seq is much more computationally demanding than microarrays,
we will be working only with a subset of data and use only one chromosome (20) to align/map the reads.
Genome, transcriptome, annotation and sample files are already downloaded, but we will show how to get them.

For you to know how to deal with raw data, we will try the RNA-seq workflow parts starting with raw reads
in FASTQ format and ending with count matrix (first two parts in the workflow above).
For that purpose we will use only four samples (two untreated and two dexamethasone treated) from the experiment
and subset each of them to 1 250 000 reads.

Then for differential expression and gene set analysis/functional profiling (last two parts in the workflow above)
we will use R objects of count matrices of the full dataset which are available in the R package
[airway](https://bioconductor.org/packages/release/data/experiment/html/airway.html).
Of course we would be able to work with the full dataset and finish with the same R objects,
but that would take a large amount of computational time.

***

# Config

First we set variables with paths to files and directories:

```{bash}
# This is a root directory for data.
BASE_DATA_DIR="../data"

# Here you will find experiment data.
EXPERIMENT_DATA_DIR="$BASE_DATA_DIR/experiment"

# A directory to store trimmed data.
TRIMMED_DATA_DIR="$BASE_DATA_DIR/experiment_trimmed"
mkdir $TRIMMED_DATA_DIR

# A directory for FastQC output.
FASTQC_DIR="fastqc"
mkdir $FASTQC_DIR

# A directory for MultiQC output. Will be created automatically by MultiQC.
MULTIQC_DIR="multiqc"

# A directory for logs.
LOG_DIR="logs"
mkdir $LOG_DIR

# A directory for temporary files.
TMP_DIR="tmp"
mkdir $TMP_DIR

# Adapter files.
TRIMMOMATIC_ADAPTERS_PATH="/usr/share/trimmomatic/TruSeq*"
ADAPTERS_FASTA_FILE="adapters.fa"
ADAPTERS_TSV_FILE="adapters.tsv"

# Array of sample names (SRA IDs).
SAMPLE_NAMES=("SRR1039508" "SRR1039509" "SRR1039512" "SRR1039513")
```

***

# Downloading experiment data from SRA

To download data from SRA, you need to use a special tool [fasterq-dump](https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump)
from a collection [SRA Tools](https://github.com/ncbi/sra-tools).

> Do not run this as data are already downloaded and you have copied the FASTQs to your personal directory.
  You can just review the code `r emo::ji("slightly_smiling_face")`

***

We will loop through sample SRA IDs and for each sample obtain two files ending with `_1.fastq` and `_2.fastq`
(because this experiment uses paired-end sequencing).
You can inspect `echo ${SAMPLE_NAMES[@]}` to see what we are actually iterating over.

```{bash}
for sample_id in ${SAMPLE_NAMES[@]}; do
  echo "Processing sample $sample_id ..."

  fasterq-dump \
    --outdir $TMP_DIR \
    --temp $TMP_DIR \
    --skip-technical \
    --threads 2 \
    --mem 8000MB \
    --log-level info \
    --progress \
    --force \
    $sample_id
done
```

***

To further reduce computing, we take only 1 250 000 reads (one read = 4 lines in FASTQ file -> 5000000 lines) from each downloaded FASTQ file:

```{bash}
for sample_id in ${SAMPLE_NAMES[@]}; do
  echo "Processing sample $sample_id ..."

  head -n 5000000 $EXPERIMENT_DATA_DIR/${sample_id}_1.fastq > $EXPERIMENT_DATA_DIR/${sample_id}_1_subset.fastq
  head -n 5000000 $EXPERIMENT_DATA_DIR/${sample_id}_2.fastq > $EXPERIMENT_DATA_DIR/${sample_id}_2_subset.fastq
done
```

Let's inspect the first read pair of one of the downloaded samples:

```{bash}
cat ../data/experiment/SRR1039508_1.fastq | head -n 4
```

```
@SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
CATTGCTGATACCAANNNNNNNNGCATTCCTCAAGGTCTTCCTCCTTCCCTTACGGAATTACA
+SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
HJJJJJJJJJJJJJJ########00?GHIJJJJJJJIJJJJJJJJJJJJJJJJJHHHFFFFFD
```

```{bash}
cat ../data/experiment/SRR1039508_2.fastq | head -n 4
```

```
@SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
CAGATGAGGCGTGTTGGCCAGAGAGCCATTGTCAACAGCAGAGATGNNNNNNNNNNNNAATCC
+SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
HJJJJJJJJJJHIIIJJJJJJJJJJJJJJJJJJJJJJJHIJIJHII#################
```

You can see that each read in FASTQ format occupies four lines:

1. Header: contains read ID, technical data from sequencer and read length.
2. Read bases.
3. Header again.
4. [Base quality](https://en.wikipedia.org/wiki/Phred_quality_score) encoded as ASCII characters.
   There are different [encoding tables](https://en.wikipedia.org/wiki/FASTQ_format#Encoding), but
   newer Illumina sequencers (1.8+) are using encoding called Phred+33.

Read pairs (forward and reverse reads) are synchronised by their positions in the corresponding files -
you can see that read IDs are same in the files above.

***

# Technical quality control - raw reads

Now we will look at the technical quality of our RNA-seq experiment.
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) is probably most used for this purpose.
On the website you can look at example reports with [good](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html), [bad](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html), and other types of data.
Very helpful is `FastQC` [documentation](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) explaining the output from each step.

`FastQC` is also checking for adapter contamination, but it needs a tab-separated values (TSV) file (format: `name[tab]sequence`) instead of FASTA.
[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) is a tool used for trimming sequencing adapters and low quality bases from reads.
We will use it later for this purpose, but for now we utilize its FASTA files of adapters and use it in the technical quality control.
We join all FASTAs to a single one and convert it to the TSV format with a simple Python script `fasta_to_table.py`.

```{bash}
cat $TRIMMOMATIC_ADAPTERS_PATH > $ADAPTERS_FASTA_FILE
echo "" >> $ADAPTERS_FASTA_FILE
head $ADAPTERS_FASTA_FILE
python3 fasta_to_table.py $ADAPTERS_FASTA_FILE > $ADAPTERS_TSV_FILE
head $ADAPTERS_TSV_FILE
```

Let's run `FastQC` now:

```{bash}
fastqc -t 2 -a $ADAPTERS_TSV_FILE -o $FASTQC_DIR $EXPERIMENT_DATA_DIR/*_subset.fastq
```

<p>
  <button class="btn btn-primary btn-sm" type="button" data-toggle="collapse" data-target="#div_info_fastqc" aria-expanded="false" aria-controls="div_info_fastqc">
    Show parameters info
  </button>
</p>

<div class="collapse div-collapse-info" id="div_info_fastqc">

- `-t 2`{.bash}
  - Number of threads to use for processing. Our VM has only 2 cores available, thus 2 threads are the maximum for optimal performance.
- `-a $ADAPTERS_TSV_FILE`{.bash}
  - Path to TSV file with adapter sequences.
- `-o $FASTQC_DIR`{.bash}
  - Output directory.
- `$EXPERIMENT_DATA_DIR/*_subset.fastq`{.bash}
  - Input FASTQ files.

</div>

`FastQC` has created report for each of input FASTQ files.
You can inspect them in the `fastqc` directory.

Another great tool is [MultiQC](https://multiqc.info/) which is able to aggregate output from multiple
tools/samples into a single beautiful report.
See [this list of supported tools](https://multiqc.info/#supported-tools) and its [documentation](https://multiqc.info/docs/#multiqc-modules).

Now we use it to aggregate the `FastQC` reports:

```{bash}
export LC_ALL=C.UTF-8
export LANG=C.UTF-8
multiqc --outdir $MULTIQC_DIR --filename "multiqc_raw.html" $FASTQC_DIR/*_subset_fastqc.*
```

<p>
  <button class="btn btn-primary btn-sm" type="button" data-toggle="collapse" data-target="#div_info_multiqc" aria-expanded="false" aria-controls="div_info_multiqc">
    Show parameters info
  </button>
</p>

<div class="collapse div-collapse-info" id="div_info_multiqc">

- `--outdir $MULTIQC_DIR`{.bash}
  - Output directory.
- `--filename "multiqc_raw.html"`{.bash}
  - Output filename.
- `$FASTQC_DIR/*_subset_fastqc.*`{.bash}
  - Files to aggregate.
  - Could be also a directory to recursively search in.
    `MultiQC` is smart and automatically recognizes from which tool does a file come from.

</div>

Now inspect the report `multiqc_raw.html` in `multiqc` directory.
Overall, data is of very high quality.

***

# Trimming the reads

Usually the first step of sequencing data preprocessing is trimming.
It removes adapter (technical) part of reads and filter/trim reads based on their [PHRED](https://en.wikipedia.org/wiki/Phred_quality_score) quality score.

However, trimming is not always needed.
**You should always closely look at technical quality control reports and then decide whether trimming is really needed.**
Also, [here](https://link.springer.com/article/10.1186/s12859-016-0956-2) (see Conclusions) and
[here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3871669/) are papers evaluating the effect of trimming on various downstream analysis.
In case of high quality data (overall good quality of base calls), you should only trim adapters (if there are any)
and don't perform trimming based on quality.

In the reports from `FastQC` (aggregated by `MultiQC`), you can see the high quality of data and missing adapter contamination.
Anyway, for the sake of exercising, we will do the trimming.

There are several popular tools for trimming:

- [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
- [cutadapt](https://cutadapt.readthedocs.io/en/stable/)
- [Trim Galore!](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) - a wrapper around `cutadapt` and `FastQC`
- [fastp](https://github.com/OpenGene/fastp) - this is actually
  "an ultra-fast all-in-one FASTQ preprocessor (QC/adapters/trimming/filtering/splitting/merging...)"
- [FASTX-Toolkit](http://hannonlab.cshl.edu/fastx_toolkit/) - similar to `fastp`

We will be using [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
([manual](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf)) which is designed for Illumina data
and which we are familiar with. You can check the evaluation of other trimming tools in the [second paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3871669/) above. But `Trimmomatic` seems to be quite good `r emo::ji("slightly_smiling_face")`

Before you run `Trimmomatic`, you should know the technical details about sequencing
(used kit, machine, read length) and also look at results from `FastQC`,
and decide how to set the parameters.

Here is a short explanation of `Trimmomatic`'s parameters:

- ILLUMINACLIP:&lt;FASTA with adapters&gt;:&lt;seed mismatches&gt;:&lt;palindrome clip threshold&gt;:&lt;simple clip threshold&gt;
  - Remove Illumina adapters and also look at adapter read-through.
- LEADING:quality
  - Cut bases off the start of a read, if below a threshold quality.
- TRAILING:quality
  - Cut bases off the end of a read, if below a threshold quality.
- SLIDINGWINDOW:window size:required quality
  - Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold.
- MINLEN:length
  - Drop the read if it is below a specified length.

To keep the pairing of reads, `Trimmomatic` will drop an entire read pair if one of the reads doesn't pass filtering.
The "good" read from such pair will be saved in a separate file (see "Show parameters info" below).

```{bash}
for sample_id in ${SAMPLE_NAMES[@]}; do
  echo "Processing sample $sample_id ..."
  
  sample_1_file="${sample_id}_1_subset.fastq"
  sample_2_file="${sample_id}_2_subset.fastq"
  
  TrimmomaticPE \
    -threads 2 \
    -phred33 \
    $EXPERIMENT_DATA_DIR/$sample_1_file \
    $EXPERIMENT_DATA_DIR/$sample_2_file \
    $TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_1P.fastq \
    $TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_1U.fastq \
    $TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_2P.fastq \
    $TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_2U.fastq \
    ILLUMINACLIP:$ADAPTERS_FASTA_FILE:2:30:8 \
    LEADING:13 \
    TRAILING:13 \
    SLIDINGWINDOW:4:19 \
    MINLEN:36 \
    2> $LOG_DIR/${sample_id}_trimmomatic.log
done
```

<p>
  <button class="btn btn-primary btn-sm" type="button" data-toggle="collapse" data-target="#div_info_trimmomatic" aria-expanded="false" aria-controls="div_info_trimmomatic">
    Show parameters info
  </button>
</p>

<div class="collapse div-collapse-info" id="div_info_trimmomatic">

- `-phred33`{.bash}
  - Base quality coding is phred33 ([info](https://en.wikipedia.org/wiki/FASTQ_format#Encoding)).
- `$EXPERIMENT_DATA_DIR/$sample_1_file"`{.bash}
  - Input forward reads.
- `$EXPERIMENT_DATA_DIR/$sample_2_file`{.bash}
  - Input reverse reads.
- `$TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_1P.fastq`{.bash}
  - Output forward reads.
- `$TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_1U.fastq`{.bash}
  - Output unpaired forward reads (their reverse mate was dropped).
- `$TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_2P.fastq`{.bash}
  - Output reverse reads.
- `$TRIMMED_DATA_DIR/${sample_id}_subset_trimmomatic_2U.fastq`{.bash}
  - Output unpaired reverse reads (their forward mate was dropped).
- `ILLUMINACLIP:$ADAPTERS_FASTA_FILE:2:30:8`{.bash}
  - Parameters for removing adapters.
- `LEADING:13`{.bash}
  - Cut bases off the start of a read, if below a threshold quality of 13.
- `TRAILING:13`{.bash}
  - Cut 13 bases off the end of a read, if below a threshold quality.
- `SLIDINGWINDOW:4:19`{.bash}
  - Sliding window of 4 bases and threshold quality of 19.
    If mean quality in the sliding windows goes under 19, read will be trimmed at this position.
- `MINLEN:36`{.bash}
  - Minimum length of read. If less than, it will be dropped.
    In our experiment, reads have the same length, so this is probably not utilized.
- `2> $LOG_DIR/trimmomatic.log`{.bash}
  - Write log to file (`Trimmomatic` is using stderr for that.).
    We will later include this log in `MultiQC` report.

</div>

<p>
  <button class="btn btn-primary btn-sm" type="button" data-toggle="collapse" data-target="#div_info_trimmomatic_log" aria-expanded="false" aria-controls="div_info_trimmomatic_log">
    Show Trimmomatic's log
  </button>
</p>

<div class="collapse div-collapse-info" id="div_info_trimmomatic_log">

Sample SRR1039508

```
TrimmomaticPE: Started with arguments:
 -threads 2 -phred33 ../data/experiment/SRR1039508_1_subset.fastq ../data/experiment/SRR1039508_2_subset.fastq ../data/experiment_trimmed/SRR1039508_subset_trimmomatic_1P.fastq ../data/experiment_trimmed/SRR1039508_subset_trimmomatic_1U.fastq ../data/experiment_trimmed/SRR1039508_subset_trimmomatic_2P.fastq ../data/experiment_trimmed/SRR1039508_subset_trimmomatic_2U.fastq ILLUMINACLIP:adapters.fa:2:30:8 LEADING:13 TRAILING:13 SLIDINGWINDOW:4:19 MINLEN:36
Using PrefixPair: 'AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT>TruSeq3_IndexedAdapterAGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'
Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'
Skipping duplicate Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'
Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT'
Using Long Clipping Sequence: 'AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAG>PrefixPE/1TACACTCTTTCCCTACACGACGCTCTTCCGATCT'
Using Long Clipping Sequence: 'AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG'
Using Long Clipping Sequence: 'TTTTTTTTTTAATGATACGGCGACCACCGAGATCTACAC'
Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT'
Using Long Clipping Sequence: 'TTTTTTTTTTCAAGCAGAAGACGGCATACGA>TruSeq2_SEAGATCGGAAGAGCTCGTATGCCGTCTTCTGCTTG'
Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC>PrefixPE/1TACACTCTTTCCCTACACGACGCTCTTCCGATCT'
Using Long Clipping Sequence: 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'
Using Long Clipping Sequence: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT'
Using Long Clipping Sequence: 'CAAGCAGAAGACGGCATACGAGATCGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT'
Using Long Clipping Sequence: 'AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT'
ILLUMINACLIP: Using 1 prefix pairs, 12 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Input Read Pairs: 1250000 Both Surviving: 1193416 (95.47%) Forward Only Surviving: 18764 (1.50%) Reverse Only Surviving: 21051 (1.68%) Dropped: 16769 (1.34%)
TrimmomaticPE: Completed successfully
```

</div>

***

# Technical quality control - trimmed reads

Now we will use `FastQC` and `MultiQC` again, on trimmed reads (only paired-end).
We also include log files from `Trimmomatic`.

```{bash}
fastqc -t 2 -a $ADAPTERS_TSV_FILE -o $FASTQC_DIR $TRIMMED_DATA_DIR/*P*.fastq
multiqc --outdir $MULTIQC_DIR --filename "multiqc_trimmed.html" $FASTQC_DIR $LOG_DIR/*_trimmomatic.log
```

Almost no difference in the technical quality, right?
As I said before, we are working with high quality data, and thus it is unlikely to see any improvement.

***
***

# HTML rendering

This chunk is not evaluated (`eval = FALSE`). Otherwise you will probably end up in recursive hell `r emo::ji("exploding_head")`

```{r, eval = FALSE, message = FALSE, warning = FALSE}
library(conflicted)
library(knitr)
library(here)

if (!require(rmdformats)) {
  BiocManager::install("rmdformats")
}

# You can set global chunk options. Options set in individual chunks will override this.
opts_chunk$set(warning = FALSE, message = FALSE, eval = FALSE)
opts_knit$set(root.dir = here())
rmarkdown::render(
  here("E07-RNA_seq/01_quality_control/01_quality_control.Rmd"),
  output_file = here("E07-RNA_seq/01_quality_control/01_quality_control.html"),
  envir = new.env()
)
```
